packages/@ember/-internals/glimmer/lib/resolver.ts

```diff
@@ -358,32 +365,28 @@ export default class RuntimeResolverImpl implements RuntimeResolver<OwnedTemplat
     const options: LookupOptions = makeOptions(moduleName, namespace);
 
     const factory =
-      owner.factoryFor(`helper:${name}`, options) || owner.factoryFor(`helper:${name}`);
+      owner.factoryFor<SimpleHelper | HelperInstance, HelperFactory<SimpleHelper | HelperInstance>>(
+        `helper:${name}`,
+        options
+      ) || owner.factoryFor(`helper:${name}`);
 
-    if (!isHelperFactory(factory)) {
+    if (factory === undefined || factory.class === undefined) {
       return null;
     }
 
-    return (args, vm) => {
-      const helper = factory.create();
-
-      if (isClassHelper(helper)) {
-        let helperDestroyable = {};
-
-        // Do this so that `destroy` gets called correctly
-        registerDestructor(helperDestroyable, () => helper.destroy(), true);
-        vm.associateDestroyable(helperDestroyable);
-      } else if (DEBUG) {
-        // Bind to null in case someone accidentally passed an unbound function
-        // in, and attempts use `this` on it.
-        //
-        // TODO: Update buildUntouchableThis to be flexible enough to provide a
-        // nice error message here.
-        helper.compute = helper.compute.bind(null);
-      }
+    const manager = getHelperManager(owner, factory.class);
 
-      return createHelperRef(helper, args.capture());
-    };
+    if (manager === undefined) {
+      return null;
+    }
+
+    if (isInternalHelperManager(manager)) {
+      return customHelper(manager, factory);
+    } else if (EMBER_GLIMMER_HELPER_MANAGER) {
+      return customHelper(manager, factory);
```

Comment:
This should be passing `factory.class` instead of `factory`, to match the other public facing APIs for managers. The manager receives the definition it was applied to, not the factory we wrap it with.

Reply:
Seems that our other public facing APIs do not agree here either. @rwjblue and I discussed, it makes more sense for these APIs to receive the definition itself and not the factory, especially taking into account the future with imports. We'll update modifier managers to receive the definition via a new capability.

Reply:
I changed it to match what we do for modifiers, but unfortunately you are right: modifier managers and component managers have different APIs in this respect.

I'll update this PR to pass the `owner.factoryFor(...).class` value in here, and also open another PR to tweak the new capability for modifier managers.

Reply:
Updated...